db/mysql.js

```diff
@@ -676,5 +676,103 @@ module.exports = function (log, error) {
     )
   }
 
+  var ACQUIRE_EVENT_LOG_LOCK = 'CALL acquireEventLogLock_1()'
+  var RELEASE_EVENT_LOG_LOCK = 'CALL releaseEventLogLock_1()'
+
+  MySql.prototype._doWithEventLogLock = function (fn) {
+    // We have to call GET_LOCK() and RELEASE_LOCK() on the same
+    // db connection or things will get very confused.
+    return this.getConnection('MASTER')
+      .then(
+        function (connection) {
+          return query(connection, ACQUIRE_EVENT_LOG_LOCK, [])
+            .then(
+              function (results) {
+                if (!results[0][0].success) {
+                  throw new Error('Failed to acquire event log lock')
+                }
+                // Execute the callback, coercing it to a promise.
+                return P.resolve()
+                  .then(
+                    function () {
+                     return fn(connection)
```

Comment:
I did this because there's no contract requiring `fn` to return a promise; there may be a neater way, it's just the first thing that came to mind.